### PR TITLE
Apply the current develop over main.

### PR DESCRIPTION
**Pull request by jjos2372**

### CODE REVIEW

**Review by frankTheTank72 — APPROVED** [review]
Let´s rock it  :-) 
Great work

**Review by frankTheTank72 — APPROVED** [review]

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -24,32 +24,43 @@ jobs:
...
+          asset_name: signum-node-${{ steps.get_version.outputs.VERSION }}.zip
+          asset_content_type: application/zip
+      - name: Upload Release Asset - Zip checksum
+        id: upload-release-asset-zip-checksum
+        uses: actions/upload-release-asset@v1.0.1
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        with:
+          upload_url: ${{ steps.create_release.outputs.upload_url }} # This pulls from the CREATE RELEASE step above, referencing it's ID to get its outputs object, which include a `upload_url`. See this blog post for more info: https://jasonet.co/posts/new-features-of-github-actions/#passing-data-to-future-steps
+          asset_path: ./build/checksums/signum-node.zip.sha256
```
Comment (ohager) [inline .github/workflows/release.yml:63]: looks very good

File: README.md [hunk 2]
```diff
@@ -131,10 +182,10 @@ Run these commands (`master` is always the latest stable release):
 ```bash
 git clone https://github.com/signum-network/signum-node.git
 cd signum-node
-mvn package
+./gradlew dist
 ```
 
-Your packaged release will now be available in the `dist` directory.
+Your packaged release will now be available in the `build/distribution` directory.
```
Comment (ohager) [inline README.md:188]: nice overhaul

File: build.gradle [hunk 3]
```diff
@@ -0,0 +1,238 @@
+plugins {
+    id 'com.github.johnrengelman.shadow' version '4.0.4'
+    id 'java'
+    id 'maven'
+    id 'application'
+    id 'edu.sc.seis.launch4j' version '2.5.1'
+    // id 'edu.sc.seis.macAppBundle' version '2.3.0'
```
Comment (ohager) [inline build.gradle:7]: ? can this be removed

File: build.gradle [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+    compile 'javax.annotation:javax.annotation-api:1.3.2'
+    
+    compile 'org.apache.httpcomponents:httpclient:4.5.13'
+
+    compile 'com.github.jiconfont:jiconfont:1.0.0'
+    compile 'com.github.jiconfont:jiconfont-swing:1.0.1'
+    compile 'com.github.jiconfont:jiconfont-font_awesome:4.7.0.1'
+
+    //logging
+    compile 'org.apache.logging.log4j:log4j-api:2.16.0'
```
Comment (ohager) [inline build.gradle:72]: !!! good - people will look into this.. but we should take the latest: 2.17.1
https://logging.apache.org/log4j/2.x/

Version changed a lot recently due to the sec issue

File: conf/node-default.properties [hunk 5]
```diff
@@ -152,27 +152,23 @@
...
-## Accept gRPC API requests.
-
-# API.V2.Server = off
-
 ## JETTY pass-through options. See documentation at
 ## https://www.eclipse.org/jetty/documentation/9.2.22.v20170531/dos-filter.html
 ## API section:
 
-# JETTY.API.DoSFilter                   = off
+# JETTY.API.DoSFilter                   = on
```
Comment (ohager) [inline conf/node-default.properties:171]: just asking: the Filters are disabled by default?

File: conf/node-default.properties [hunk 6]
```diff
@@ -341,50 +332,13 @@
...
-# DEV.TimeWarp = 1
-
-## Peers used for testnet only.
-
-# DEV.P2P.BootstrapPeers = nivbox.co.uk; 77.66.65.240; testnetwallet.burstcoin.ro; 2.29.196.249; 176.9.197.183;
+## Run with a different network
 
-## Testnet only. These peers will always be sent rebroadcast transactions. They are also automatically added to DEV.P2P.BootstrapPeers, so no need for duplicates.
+# Testnet network
+node.network = signum.net.TestnetNetwork
```
Comment (ohager) [inline conf/node-default.properties:338]: So, initial default conf is for testnet work?

File: resources/db/migration_mariadb/V6_1__indirect_upgrade.sql [hunk 7]
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE indirect_incoming ADD amount bigint(20) DEFAULT '0';
```
Comment (ohager) [inline resources/db/migration_mariadb/V6_1__indirect_upgrade.sql:1]: what's the difference between bigint(20) and BigInt (like in V6_1_indirect_upgrade.sql) - could this cause inconsistencies?

File: src/brs/BurstGUI.java [hunk 8]
```diff
@@ -224,35 +231,45 @@ private TrayIcon createTrayIcon() {
...
     	MenuItem shutdownItem = new MenuItem("Shutdown the node");
 
     	JButton openPhoenixButton = new JButton(openPheonixWalletItem.getLabel(), IconFontSwing.buildIcon(FontAwesome.FIRE, 18, iconColor));
-    	JButton openClassicButton = new JButton(openClassicWalletItem.getLabel(), IconFontSwing.buildIcon(FontAwesome.WINDOW_RESTORE, 18, iconColor));
+        JButton openClassicButton = new JButton(openClassicWalletItem.getLabel(), IconFontSwing.buildIcon(FontAwesome.WINDOW_RESTORE, 18, iconColor));
+        JButton openApiButton = new JButton(openApiItem.getLabel(), IconFontSwing.buildIcon(FontAwesome.BOOK, 18, iconColor));
     	JButton editConfButton = new JButton("Edit conf file", IconFontSwing.buildIcon(FontAwesome.PENCIL, 18, iconColor));
         JButton popOff10Button = new JButton("Pop off 10 blocks", IconFontSwing.buildIcon(FontAwesome.STEP_BACKWARD, 18, iconColor));
         JButton popOff100Button = new JButton("Pop off 100 blocks", IconFontSwing.buildIcon(FontAwesome.BACKWARD, 18, iconColor));
         // TODO: find a way to actually store permanently the max block available to pop-off, otherwise we can break it
```
Comment (ohager) [inline src/brs/BurstGUI.java:244]: Is it still relevant? Otherwise, remove it

File: src/brs/BurstGUI.java [hunk 9]
```diff
@@ -224,35 +231,45 @@ private TrayIcon createTrayIcon() {
...
+    	toolBar.add(openApiButton);
+    	if(Burst.getPropertyService().getBoolean(Props.EXPERIMENTAL)) {
           toolBar.add(popOff10Button);
           toolBar.add(popOff100Button);
 //          toolBar.add(popOffMaxButton);
     	}
 
-    	openPheonixWalletItem.addActionListener(e -> openWebUi(false));
-        openClassicWalletItem.addActionListener(e -> openWebUi(true));
+    	openPheonixWalletItem.addActionListener(e -> openWebUi("/phoenix"));
```
Comment (ohager) [inline src/brs/BurstGUI.java:271]: minor typo in `openPheonix` -> `openPhoenix`

File: src/brs/GeneratorImpl.java [hunk 10]
```diff
@@ -162,20 +157,26 @@ public double getCommitmentFactor(long commitment, long averageCommitment, int b
   @Override
   public BigInteger calculateDeadline(BigInteger hit, long capacityBaseTarget, long commitment, long averageCommitment, int blockHeight) {
     BigInteger deadline = hit.divide(BigInteger.valueOf(capacityBaseTarget));
+    
+    double blockTime = fluxCapacitor.getValue(FluxValues.BLOCK_TIME);
+    double lnScale = (blockTime) / Math.log(blockTime);
+
     if(fluxCapacitor.getValue(FluxValues.POC_PLUS, blockHeight)) {
+      // private static final double lnScale = 49d; // value that would keep the legacy network size estimation close to real capacity
```
Comment (ohager) [inline src/brs/GeneratorImpl.java:165]: if we don't need it, we can remove it

File: src/brs/db/sql/SqlIndirectIncomingStore.java [hunk 11]
```diff
@@ -44,25 +54,29 @@ void save(DSLContext ctx, IndirectIncoming indirectIncoming) {
...
             @Override
-            void save(DSLContext ctx, IndirectIncoming[] indirectIncomings) {
+            void save(DSLContext ctx, Collection<IndirectIncoming> indirectIncomings) {
+              Iterator<IndirectIncoming> iterator = indirectIncomings.iterator();
+              while(iterator.hasNext()) {
+                // break into batches of 50k queries max
                 List<Query> queries = new ArrayList<>();
-                for (IndirectIncoming indirectIncoming: indirectIncomings) {
-                    queries.add(getQuery(ctx, indirectIncoming));
+                for (int i = 0; i < 50000 && iterator.hasNext(); i++) {
```
Comment (ohager) [inline src/brs/db/sql/SqlIndirectIncomingStore.java:62]: is this chunking a performance improvement?

File: src/brs/db/sql/SqlSubscriptionStore.java [hunk 12]
```diff
@@ -88,13 +89,11 @@ private static Condition getUpdateOnBlockClause(final int timestamp) {
     return subscriptionTable.getManyBy(getUpdateOnBlockClause(timestamp), 0, -1);
   }
 
-  private void saveSubscription(DSLContext ctx, Subscription subscription) {
-    ctx.mergeInto(SUBSCRIPTION, SUBSCRIPTION.ID, SUBSCRIPTION.SENDER_ID, SUBSCRIPTION.RECIPIENT_ID, SUBSCRIPTION.AMOUNT, SUBSCRIPTION.FREQUENCY, SUBSCRIPTION.TIME_NEXT, SUBSCRIPTION.HEIGHT, SUBSCRIPTION.LATEST)
-            .key(SUBSCRIPTION.ID, SUBSCRIPTION.SENDER_ID, SUBSCRIPTION.RECIPIENT_ID, SUBSCRIPTION.AMOUNT, SUBSCRIPTION.FREQUENCY, SUBSCRIPTION.TIME_NEXT, SUBSCRIPTION.HEIGHT, SUBSCRIPTION.LATEST)
-            .values(subscription.id, subscription.senderId, subscription.recipientId, subscription.amountNQT, subscription.frequency, subscription.getTimeNext(), Burst.getBlockchain().getHeight(), true)
-            .execute();
+  private Query insertSubscription(DSLContext ctx, Subscription subscription) {
+    return ctx.insertInto(SUBSCRIPTION, SUBSCRIPTION.ID, SUBSCRIPTION.SENDER_ID, SUBSCRIPTION.RECIPIENT_ID, SUBSCRIPTION.AMOUNT, SUBSCRIPTION.FREQUENCY, SUBSCRIPTION.TIME_NEXT, SUBSCRIPTION.HEIGHT, SUBSCRIPTION.LATEST)
```
Comment (ohager) [inline src/brs/db/sql/SqlSubscriptionStore.java:93]: :+1:

File: src/brs/db/sql/migration/V6_4__FixBurntAmount.java [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+              while (txResult.next()) {
+                byte[] attachment_bytes = txResult.getBytes(1);
+                byte subtype = txResult.getByte(2);
+                byte version = txResult.getByte(3);
+                long amountNQT = txResult.getLong(4);
+                
+                ByteBuffer buffer = ByteBuffer.wrap(attachment_bytes);
+                buffer.order(ByteOrder.LITTLE_ENDIAN);
+                
+                if(subtype == 1) {
```
Comment (ohager) [inline src/brs/db/sql/migration/V6_4__FixBurntAmount.java:58]: wouldn't it be better to have/use constants for the subtypes?

File: src/brs/fluxcapacitor/FluxCapacitorImpl.java [hunk 14]
```diff
@@ -13,8 +11,8 @@
     private final PropertyService propertyService;
     private final Blockchain blockchain;
 
-    // Map of Flux Value -> Change Height -> Index of ValueChange in FluxValue. Used as a cache.
-    private final Map<FluxValue<?>, Map<Integer, Integer>> valueChangesPerFluxValue = new ConcurrentHashMap<>();
+    // Map used as a cache.
```
Comment (ohager) [inline src/brs/fluxcapacitor/FluxCapacitorImpl.java:14]: comment not necessary

File: src/brs/fluxcapacitor/FluxValues.java [hunk 15]
```diff
@@ -19,8 +19,13 @@ private FluxValues() {
     public static final FluxEnable SIGNUM = new FluxEnable(HistoricalMoments.SIGNUM);
     public static final FluxEnable POC_PLUS = new FluxEnable(HistoricalMoments.POC_PLUS);
     public static final FluxEnable SPEEDWAY = new FluxEnable(HistoricalMoments.SPEEDWAY);
+    public static final FluxEnable SMART_TOKEN = new FluxEnable(HistoricalMoments.SMART_TOKEN);
```
Comment (ohager) [inline src/brs/fluxcapacitor/FluxValues.java:22]: :partying_face:

File: src/brs/fluxcapacitor/HistoricalMoments.java [hunk 16]
```diff
@@ -3,40 +3,35 @@
...
+    public static HistoricalMoments AT_FIX_BLOCK_2 = new HistoricalMoments(67000, Props.AT_FIX_BLOCK_2_BLOCK_HEIGHT);
+    public static HistoricalMoments AT_FIX_BLOCK_3 = new HistoricalMoments(92000, Props.AT_FIX_BLOCK_3_BLOCK_HEIGHT);
+    public static HistoricalMoments AT_FIX_BLOCK_4 = new HistoricalMoments(255000, Props.AT_FIX_BLOCK_4_BLOCK_HEIGHT);
+    public static HistoricalMoments PRE_POC2 = new HistoricalMoments(500000, Props.PRE_POC2_BLOCK_HEIGHT);
+    public static HistoricalMoments POC2 = new HistoricalMoments(502000, Props.POC2_BLOCK_HEIGHT);
+    public static HistoricalMoments SODIUM = new HistoricalMoments(765_000, Props.SODIUM_BLOCK_HEIGHT);
+    public static HistoricalMoments SIGNUM = new HistoricalMoments(875_500, Props.SIGNUM_HEIGHT);
+    public static HistoricalMoments POC_PLUS = new HistoricalMoments(878_000, Props.POC_PLUS_HEIGHT);
+    public static HistoricalMoments SPEEDWAY = new HistoricalMoments(941_100, Props.SPEEDWAY_HEIGHT);
+    public static HistoricalMoments SMART_TOKEN = new HistoricalMoments(Integer.MAX_VALUE, Props.SMART_TOKEN_HEIGHT);
```
Comment (ohager) [inline src/brs/fluxcapacitor/HistoricalMoments.java:20]: not defined yet :grin:

File: src/brs/http/API.java [hunk 17]
```diff
@@ -227,9 +222,9 @@ private void letsencryptToPkcs12(String letsencryptPath, String p12File, String
     process.waitFor();
   }
 
-  private String regexpEscapeUrl(String url) {
-    return url.replace("/", "\\/");
-  }
+//  private String regexpEscapeUrl(String url) {
```
Comment (ohager) [inline src/brs/http/API.java:225]: ? remove it?

File: src/brs/http/API.java [hunk 18]
```diff
@@ -241,18 +236,18 @@ public void shutdown() {
...
-      super(regex, replacement);
-      this.baseDirectory = baseDirectory;
-    }
-
-    @Override
-    public String apply(String target, HttpServletRequest request, HttpServletResponse response, Matcher matcher) throws IOException {
-      return new File(baseDirectory, target).exists() ? target : super.apply(target, request, response, matcher);
-    }
-  }
+//  private static class RegexOrExistsRewriteRule extends RewriteRegexRule {
```
Comment (ohager) [inline src/brs/http/API.java:239]: same

File: src/brs/http/APITestServlet.java [hunk 19]
```diff
@@ -268,8 +277,8 @@ private String form(String requestType, boolean singleView, String className, Li
       buf.append("\" target=\"_blank\" style=\"font-weight:normal;font-size:14px;color:#777;\"><span class=\"glyphicon glyphicon-new-window\"></span></a>");
       buf.append(" &nbsp;&nbsp;");
     }
-    buf.append("<a style=\"font-weight:normal;font-size:14px;color:#777;\" href=\"/doc/");
-    buf.append(className.replace('.','/')).append(".html\" target=\"_blank\">javadoc</a>");
+    //buf.append("<a style=\"font-weight:normal;font-size:14px;color:#777;\" href=\"/doc/");
```
Comment (ohager) [inline src/brs/http/APITestServlet.java:280]: remove what we dont need :wink:

**Review by ohager — APPROVED** [review]
Wow... big refactoring, nice cleanups (e.g. grpc). I did not understand all the implementation details of the business rules. I saw several commented code, which should be removed.

File: build.gradle (hunk 3 above)
Comment (jjos2372) [reply build.gradle:7]: removed.

File: build.gradle (hunk 4 above)
Comment (jjos2372) [reply build.gradle:72]: Updated to 2.17.1

File: conf/node-default.properties (hunk 5 above)
Comment (jjos2372) [reply conf/node-default.properties:171]: they are now enabled by default and localhost is whitelisted

File: src/brs/http/APITestServlet.java (hunk 19 above)
Comment (jjos2372) [reply src/brs/http/APITestServlet.java:280]: done

File: src/brs/http/API.java (hunk 18 above)
Comment (jjos2372) [reply src/brs/http/API.java]: removed

**Pull request merged**